core/src/main/java/com/microsoft/applicationinsights/channel/concrete/ATelemetryChannel.java

```diff
@@ -0,0 +1,390 @@
+     */
+    public ATelemetryChannel(Map<String, String> namesAndValues) {
+        boolean developerMode = false;
+        String endpointAddress = null;
+        int maxInstantRetries = DEFAULT_MAX_INSTANT_RETRY;
+
+        LimitsEnforcer maxTelemetryBufferCapacityEnforcer = createDefaultMaxTelemetryBufferCapacityEnforcer(null);
+
+        LimitsEnforcer sendIntervalInSecondsEnforcer = createDefaultSendIntervalInSecondsEnforcer(null);
+
+        boolean throttling = true;
+        if (namesAndValues != null) {
+            throttling = Boolean.valueOf(namesAndValues.get(THROTTLING_ENABLED_NAME));
+            developerMode = Boolean.valueOf(namesAndValues.get(DEVELOPER_MODE_NAME));
+            try {
+                String instantRetryValue = namesAndValues.get(INSTANT_RETRY_NAME);
+                if (instantRetryValue != null) {
+                    maxInstantRetries = Integer.parseInt(instantRetryValue);
+                }
+
+            } catch (NumberFormatException e) {
+                InternalLogger.INSTANCE.error("Unable to parse configuration setting %s to integer value.%nStack Trace:%n%s", INSTANT_RETRY_NAME, ExceptionUtils.getStackTrace(e));
+            }
+
+            if (!developerMode) {
+                developerMode = Boolean.valueOf(System.getProperty(DEVELOPER_MODE_SYSTEM_PROPRETY_NAME));
+            }
+            endpointAddress = namesAndValues.get(ENDPOINT_ADDRESS_NAME);
+
+            maxTelemetryBufferCapacityEnforcer
+                    .normalizeStringValue(namesAndValues.get(MAX_TELEMETRY_BUFFER_CAPACITY_NAME));
+            sendIntervalInSecondsEnforcer
+                    .normalizeStringValue(namesAndValues.get(FLUSH_BUFFER_TIMEOUT_IN_SECONDS_NAME));
+        }
+
+        String maxTransmissionStorageCapacity =
+                namesAndValues.get(MAX_TRANSMISSION_STORAGE_CAPACITY_NAME);
+        initialize(
+                endpointAddress,
+                maxTransmissionStorageCapacity,
+                developerMode,
+                maxTelemetryBufferCapacityEnforcer,
+                sendIntervalInSecondsEnforcer,
+                throttling,
+                maxInstantRetries);
+    }
+
+    protected synchronized void initialize(String endpointAddress, String maxTransmissionStorageCapacity,
+                                       boolean developerMode, LimitsEnforcer maxTelemetryBufferCapacityEnforcer,
+                                       LimitsEnforcer sendIntervalInSeconds, boolean throttling, int maxInstantRetry) {
+        makeSureEndpointAddressIsValid(endpointAddress);
+
+
+        telemetriesTransmitter = getTransmitterFactory().create(endpointAddress, maxTransmissionStorageCapacity, throttling, maxInstantRetry);
+        telemetryBuffer = new TelemetryBuffer<>(telemetriesTransmitter, maxTelemetryBufferCapacityEnforcer, sendIntervalInSeconds);
+
+        setDeveloperMode(developerMode);
+    }
+
+    protected synchronized TransmitterFactory<T> getTransmitterFactory() {

```

Comment:
Is this expected to be a high-volume call? I don't know how you guys do it in Java, but over in .NET land we usually do a null check, then acquire the lock, and then do another null check. That way you don't take a lock every time someone calls this method. But of course if this is only called once or twice that wouldn't matter.

Reply:
No, this is just called on init.